### ISSUE TITLE
Request: option to automatically switch to REPL after racket-run

### ISSUE THREAD

**Issue by jjackaman** [greghendershott/racket-mode#102]
In DrRacket, when you click RUN, the definitions area is evaluated and the cursor is automatically switched to the interactions area (awaiting input at the REPL),  I'm going through HtDP2e and 90% of the time I have to manually switch to the REPL when pressing F5 in racket-mode, it would be nice to have a option in racket-mode to allow point to automatically switch to the REPL after racket-run.

**Comment by greghendershott** [comment]
Personally I love the way it works now, because my workflow is mostly editing code and running, and I don't mind typing <kbd>C-x o</kbd> when I do want to explore in the REPL.  But I understand you have a different workflow. I agree an option would be good.

Meanwhile I think you could put this in your emacs init file:

```
(eval-after-load 'racket-mode
  '(defadvice racket-run (after switch-to-racket-repl activate)
     "After `racket-run', make the REPL window current."
     (select-window (get-buffer-window racket--repl-buffer-name))))
```

Just keep in mind that the Racket REPL window currently doesn't know about `racket-run` and the current file being evaluated. So you would need to <kbd>C-x o</kbd> to get back to your .rkt file buffer, to `racket-run` again.

---

Notes to self:

- Add a `,path` command to the REPL, so that `racket-repl-mode` can learn what file (if any) is active. (It can't simply remember this from the last `racket-run` command, because the user could have typed a `,run` or `,top` command at the REPL prompt. It has to ask the REPL itself, with the proposed command `,path`.)

- Add the traditional <kbd>C-c C-z</kbd> commands to both `racket-mode` and `racket-repl-mode`, to switch from one to the other. (The former could be done already, but the latter requires the previous bullet point.)

- Add an option for `racket-run` effectively to do a <kbd>C-c C-z</kbd>, to select the REPL window. Close this issue. :)

**Comment by BartAdv** [comment]
Maybe all functions that send stuff to REPL for evaluation should have additional argument denoting whether to switch buffer after evaluation?

I'm just investigating spacemacs, and there is a convention for keybindings for languages that interact with REPLs. This convention assumes every REPL interaction can come in two flavours, with and without switching to REPL buffer.

**Comment by greghendershott** [comment]
@BartAdv 

> I'm just investigating spacemacs, and there is a convention for keybindings for languages that interact with REPLs. This convention assumes every REPL interaction can come in two flavours, with and without switching to REPL buffer.

I skimmed the spacemacs docs just now and couldn't find this keybinding convention. What is it?

---

BTW there seems to be a third approach for Lispy languages in Emacs; Not only is the REPL not selected, it's not even shown. Instead, the result is shown in the echo area. For example that's what happens with <kbd>C-x C-e</kbd> in an Emacs-Lisp or Clojure buffer.

**Comment by BartAdv** [comment]
https://github.com/syl20bnr/spacemacs/blob/master/doc/CONVENTIONS.md#repls

But I've figured out `racket-mode` philosophy of being similar to
`DrRacket` means there just might be no need for other types of REPL
evaluation.

**Comment by syl20bnr** [comment]
@greghendershott the conventions are stored in this file: https://github.com/syl20bnr/spacemacs/blob/master/doc/CONVENTIONS.md

`evaluation` section is for results printed in the echo area, `REPLs send code` section is for sending code to a REPL buffer.

The original idea for these conventions come from ESS which has a lot of REPL interaction functions. Often the `send code and switch to REPL` does not exist (i.e. in python layer) so we define additional functions in spacemacs to handle such cases.

**Comment by greghendershott** [comment]
@BartAdv @syl20bnr Thanks for the links.

**Comment by greghendershott** [comment]
Originally <kbd>F5</kbd> was bound to `racket-run`, which does not switch to the REPL. (<kbd>F5</kbd> is the DrRacket key. It is a very un-Emacs-y key. Technically it's reserved for the user.)

About a month ago I also bound <kbd>C-c C-k</kbd> to `racket-run`, because I grew accustomed to using that in some other modes like Clojure mode. And it's more Emacs-y.

For this issue, what I'm trying now (not yet committed and pushed):

- Leave <kbd>C-c C-k</kbd> bound to `racket-run` (which still doesn't switch to the REPL).
- Bind <kbd>C-c C-z</kbd> to a new `racket-switch-to-repl` command.
- Bind <kbd>F5</kbd> to a new `racket-run-and-switch-to-repl` command, which is simply the composition of the previous two.

And with these commands exposed, people can change the default key bindings to whatever they might prefer instead.

Sound good?

**Comment by BartAdv** [comment]
Yup, from what I saw in other packages it seems to be standard way to
provide two variants: one that switch to repl, and other that does not.

**Comment by greghendershott** [comment]
Closed with commit d170577.

**Comment by greghendershott** [comment]
@jjackaman If you're installing from MELPA, it may take a few hours for it to refresh there. You'll want a version starting with "20150303".
